pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,396 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	importsource "github.com/grafana/agent/pkg/flow/internal/import-source"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+	"github.com/grafana/agent/pkg/flow/tracing"
+	"github.com/grafana/river/ast"
+	"github.com/grafana/river/parser"
+	"github.com/grafana/river/vm"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/atomic"
+)
+
+type ImportConfigNode struct {
+	id                        ComponentID
+	label                     string
+	nodeID                    string
+	componentName             string
+	globalID                  string
+	globals                   ComponentGlobals // Need a copy of the globals to create other import nodes.
+	source                    importsource.ImportSource
+	registry                  *prometheus.Registry
+	importedDeclares          map[string]string
+	importConfigNodesChildren map[string]*ImportConfigNode
+	OnComponentUpdate         func(cn NodeWithDependants)
+	logger                    log.Logger
+	inContentUpdate           bool
+
+	mut                sync.RWMutex
+	importedContentMut sync.RWMutex
+	block              *ast.BlockStmt // Current River blocks to derive config from
+	lastUpdateTime     atomic.Time
+
+	healthMut  sync.RWMutex
+	evalHealth component.Health // Health of the last evaluate
+	runHealth  component.Health // Health of running the component
+}
+
+var _ NodeWithDependants = (*ImportConfigNode)(nil)
+var _ RunnableNode = (*ImportConfigNode)(nil)
+var _ ComponentNode = (*ImportConfigNode)(nil)
```

Comment:
It feels off to me that ImportConfigNode implements the ComponentNode interface; it's not a component.

Reply:
A ComponentNode is not a component, it's a node that manages a component. ImportConfigNode manages a component to import the content

Reply:
I will rename the ComponentNode interface implemented by NativeComponentNode / DeclareComponentNode / ImportConfigNode to something else (don't know what yet) and I will create a new interface ComponentNode that is only implemented by NativeComponentNode and DeclareComponentNode

Reply:
> A ComponentNode is not a component

I'm sorry, but this isn't true. The documentation of a ComponentNode states:

> ComponentNode is a generic representation of a Flow component. 

The original intent was that ComponentNodes are components in the graph, wrapped by some node-specific logic. A node can create and manage an instance of a component, but unless it's _acting_ as a component itself, I don't think it should implement ComponentNode.

Reply:
Following up on this: I'm wrong; I had a branch for experimenting with plugins where I _also_ created a ComponentNode interface (with the definition above), and I got confused about what's in main. I even looked at my local code, but I was still checked out to that experimental branch :facepalm: Sorry for the confusion here.

In any case, I think the definition I had come up with makes logical sense: something in the DAG which implements an interface used by component is intending to act as a component from the perspective of a DAG.

I'd be OK with finding a more appropriate name or restricting the interface to that definition.